### PR TITLE
Add the url module and the tests for it

### PR DESCRIPTION
**Pull request by pmarkee**
IoT.js-DCO-1.0-Signed-off-by: Peter Marki marpeter@inf.u-szeged.hu

### CODE REVIEW

**Comment by daeyeon** [comment]
This is seemingly a preliminary work to implement https module and related modules. IMHO, parsing protocol, host, (auth,) and port from url is necessary for that. However, whether we should consider the others much is a question mark.

It would be better to get this module to have the minimum required functionalities, reduce the size as many as possible, and use it only internally. Can you clarify if there are parts we could reduce?

**Comment by pmarkee** [comment]
Do you mean we only need protocol, auth, host and port? We can definitely do that

**Comment by daeyeon** [comment]
I think that we need to support passing an url string into `http.request` or `https.request`. If we can achieve that through any other way such as using a simple regular expression, then it would be better than this url module.

FYI: https://mathiasbynens.be/demo/url-regex

**Comment by pmarkee** [comment]
I just updated the PR, I made the module much shorter and more basic. I think a regex could get quite complicated and it could be harder to change it later if it needs a change. This module could easily be used to support passing url strings to http.request and https.request IMO.

**Comment by pmarkee** [comment]
The rpi2 binary size with the module is about 2,5 kbyte larger than before

**Comment by yichoi** [comment]
Let's have more discussion about this. I am wondering whether we need to add "internal" module or not.

File: src/js/url.js [hunk 1]
```diff
@@ -0,0 +1,256 @@
...
+      case 60:
+      case 62:
+      case 92:
+      case 94:
+      case 96:
+      case 123:
+      case 124:
+      case 125:
+        // Characters that are never ever allowed in a hostname from RFC 2396
+        throw new TypeError('Invalid URL: prohibited character in the host');
```
Comment (daeyeon) [inline src/js/url.js]: `dns` module will validate a passing hostname. Do we need to check if the given the hostname string meets the specification here?

**Comment by yichoi** [comment]
please move this PR to https://github.com/Samsung/iotjs-modules

**Comment by yichoi** [comment]
closing since https://github.com/Samsung/iotjs-modules/pull/1 has landed

**Pull request closed without merging**